File: trunk/incubators/net.sf.j2s.java.org.eclipse.osgi/src/org/eclipse/core/runtime/internal/adaptor/Locker_JavaNio.java
/*******************************************************************************
 * Copyright (c) 2004 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.core.runtime.internal.adaptor;

import java.io.*;
import java.nio.channels.FileLock;
import org.eclipse.core.runtime.adaptor.EclipseAdaptor;
import org.eclipse.core.runtime.adaptor.EclipseAdaptorMsg;
import org.eclipse.osgi.framework.adaptor.FrameworkAdaptor;
import org.eclipse.osgi.framework.log.FrameworkLogEntry;
import org.eclipse.osgi.util.NLS;

/**
 * Internal class.
 */
public class Locker_JavaNio implements Locker {
	private File lockFile;
	private FileLock fileLock;
	private FileOutputStream fileStream;

	public Locker_JavaNio(File lockFile) {
		this.lockFile = lockFile;
	}

	public synchronized boolean lock() throws IOException {
		fileStream = new FileOutputStream(lockFile, true);
		try {
			fileLock = fileStream.getChannel().tryLock();
		} catch (IOException ioe) {
			// log the original exception if debugging
			if (BasicLocation.DEBUG) {
				String basicMessage = NLS.bind(EclipseAdaptorMsg.location_cannotLock, lockFile);
				FrameworkLogEntry basicEntry = new FrameworkLogEntry(FrameworkAdaptor.FRAMEWORK_SYMBOLICNAME, basicMessage, 0, ioe, null);
				EclipseAdaptor.getDefault().getFrameworkLog().log(basicEntry);
			}
			// produce a more specific message for clients
			String specificMessage = NLS.bind(EclipseAdaptorMsg.location_cannotLockNIO, new Object[] {lockFile, ioe.getMessage(), "\"-D" + BasicLocation.PROP_OSGI_LOCKING + "=none\""}); //$NON-NLS-1$ //$NON-NLS-2$
			throw new IOException(specificMessage);
		}
		if (fileLock != null)
			return true;
		fileStream.close();
		fileStream = null;
		return false;
	}

	public synchronized void release() {
		if (fileLock != null) {
			try {
				fileLock.release();
			} catch (IOException e) {
				//don't complain, we're making a best effort to clean up
			}
			fileLock = null;
		}
		if (fileStream != null) {
			try {
				fileStream.close();
			} catch (IOException e) {
				//don't complain, we're making a best effort to clean up
			}
			fileStream = null;
		}
	}
}

File: trunk/incubator/net.sf.j2s.test.mauve/j2slib/org/eclipse/core/runtime/internal/adaptor/Locker_JavaNio.js
Clazz.declarePackage ("org.eclipse.core.runtime.internal.adaptor");
Clazz.load (["org.eclipse.core.runtime.internal.adaptor.Locker"], "org.eclipse.core.runtime.internal.adaptor.Locker_JavaNio", ["java.io.FileOutputStream", "$.IOException", "org.eclipse.core.runtime.adaptor.EclipseAdaptor", "$.EclipseAdaptorMsg", "org.eclipse.core.runtime.internal.adaptor.BasicLocation", "org.eclipse.osgi.framework.log.FrameworkLogEntry", "org.eclipse.osgi.util.NLS"], function () {
c$ = Clazz.decorateAsClass (function () {
this.lockFile = null;
this.fileLock = null;
this.fileStream = null;
Clazz.instantialize (this, arguments);
}, org.eclipse.core.runtime.internal.adaptor, "Locker_JavaNio", null, org.eclipse.core.runtime.internal.adaptor.Locker);
Clazz.makeConstructor (c$, 
function (lockFile) {
this.lockFile = lockFile;
}, "java.io.File");
Clazz.overrideMethod (c$, "lock", 
function () {
this.fileStream =  new java.io.FileOutputStream (this.lockFile, true);
try {
this.fileLock = this.fileStream.getChannel ().tryLock ();
} catch (ioe) {
if (Clazz.instanceOf (ioe, java.io.IOException)) {
if (org.eclipse.core.runtime.internal.adaptor.BasicLocation.DEBUG) {
var basicMessage = org.eclipse.osgi.util.NLS.bind (org.eclipse.core.runtime.adaptor.EclipseAdaptorMsg.location_cannotLock, this.lockFile);
var basicEntry =  new org.eclipse.osgi.framework.log.FrameworkLogEntry ("org.eclipse.osgi", basicMessage, 0, ioe, null);
org.eclipse.core.runtime.adaptor.EclipseAdaptor.getDefault ().getFrameworkLog ().log (basicEntry);
}var specificMessage = org.eclipse.osgi.util.NLS.bind (org.eclipse.core.runtime.adaptor.EclipseAdaptorMsg.location_cannotLockNIO, [this.lockFile, ioe.getMessage (), "\"-Dosgi.locking=none\""]);
throw  new java.io.IOException (specificMessage);
} else {
throw ioe;
}
}
if (this.fileLock != null) return true;
this.fileStream.close ();
this.fileStream = null;
return false;
});
Clazz.overrideMethod (c$, "release", 
function () {
if (this.fileLock != null) {
try {
this.fileLock.release ();
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
} else {
throw e;
}
}
this.fileLock = null;
}if (this.fileStream != null) {
try {
this.fileStream.close ();
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
} else {
throw e;
}
}
this.fileStream = null;
}});
});
